### DISCUSSION TITLE
Full response function with part of a side on a free BC

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#28398] [Q&A General]
Hi!

I'm working on the full response function of a block of ice where I'd like to keep part of one side free while the rest of that side is on a null Dirichlet BC. I've been generating a mesh on the fly, and then using `SubdomainBoundingBoxGenerator` to target the specific area I'm interested in and further `SideSetsAroundSubdomainGenerator` to apply a BC to this zone. It works just fine when I apply a non-null Dirichlet BC on that small zone, but as soon as I was to keep it free with the rest of the side on a null Dirichlet, it's keeping it to a null Dirichlet too instead of letting it move. Do you know how I could prevent the `decoupling_bottom` boundary to be set to 0 instead of staying free? I've been searching the discussions and the doc but couldn't find a fix.
The input file can be found [here](https://github.com/AdrienWehrle/diuca/blob/PR_resonance/problems/elastic_deformation/frequency_domain_analysis/icesheet_frequency_response_functio_decoupling_zone.i)

Thank you a lot in advance for your help! :)

**Comment by GiudGiud** [comment]
Hello

Looks like the displacement BCs are being controlled by this function
```
  [bed_coupling_function]
    type = ParsedFunction
    expression = '${bed_coupled} = 0'
  []
[]
```
in this Control
```
  [bed_not_coupled]
    type = ConditionalFunctionEnableControl
    conditional_function = bed_coupling_function
    disable_objects = 'BCs::dirichlet_decoupling_bottom_x BCs::dirichlet_decoupling_bottom_y BCs::dirichlet_decoupling_bottom_z'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]
```

this function is always '0 = 0' which I m not sure is even a valid function.
Maybe try to have
expression = 'if(${bed_coupled}>-1e-6, 1, 0)'
so we dont need to wonder about the parsing rounding

**Reply by wehrad** [reply]
That's only because `bed_coupled` is a boolean given at the top of the input file, to be able to turn it on and off on a simulation basis. Then part of the bed is coupled or not during the whole simulation. And it's not function of any variable, and is modified by the user [here](https://github.com/AdrienWehrle/diuca/blob/efb49ab6173b20747b56d738db23ae3c007f08d6/problems/elastic_deformation/frequency_domain_analysis/icesheet_frequency_response_functio_decoupling_zone.i#L4).
So that if `bed_coupled` is False then the three `dirchlet_decoupling_bottom` BCs (x, y, z) are being disabled.

**Reply by GiudGiud** [reply]
> So that if bed_coupled is False then the three dirchlet_decoupling_bottom BCs (x, y, z) are being disabled.

and this part does not work currently right? They still seem to be set?
if you comment out the BC, is the displacement still blocked to 0?

**Comment by wehrad** [comment]
> and this part does not work currently right? They still seem to be set?

yes, somehow as if the new boundary created with 'decoupling_bottom` and the explicit `replace=true` is actually not truly replacing it... If I set a non-null BC on this boundary, it takes over the null one, but if I just want to keep it free, it's stuck on the null BC set around it... 

> if you comment out the BC, is the displacement still blocked to 0?

exactly, as if the boundary `decoupling_bottom` was added on top of the `bottom` that's on the entire face, but not actually replacing it.

**Reply by GiudGiud** [reply]
Sidesets don't replace other sidesets that pre exist. 

Edit you did set replace=true.

What does paraview show? It shows the sidesets 

You can use a ParsedGenerateSideset to cut one sideset out from another

**Reply by wehrad** [reply]
This is `decoupling_botom`:
![image](https://github.com/user-attachments/assets/9f090409-f5c3-4c0c-b589-7dd863e58478)

And this is `bottom`:
![image](https://github.com/user-attachments/assets/e60ae113-4d9a-4a35-bb20-02464ea55260)

So it definitely looks like it's not being replaced.. 

Maybe the issue comes from the nodesets? This feels odd:
![image](https://github.com/user-attachments/assets/641cb7bb-333e-4d9e-8cd8-fc398dfd121a)

Node sets are being created on the `decoupling_bottom` side but they're called `nodelist_6`. I'm wondering if that has anything to do with my issue...

**Reply by GiudGiud** [reply]
Nodesets do get converted back to sidesets by default

And some BCs like displacements are applied on nodesets not sidesets

So yes it is likely tied to nodesets

**Reply by wehrad** [reply]
But so the `decoupling_bottom` sideset should be tied to `nodelist_6`.. And that's the case since when I set a non-null BC then it works fine. Do you mean that when no displacement BC is set on `decoupling_bottom` then the nodeset associated to this boundary (`nodelist_6`) is being removed for the `bottom` nodeset to take over?

**Comment by wehrad** [comment]
> You can use a ParsedGenerateSideset to cut one sideset out from another

Ok I'll try that. I decided not to initially since this involves hard coding geometry dimensions rather than calling a boundary. It would work here with a flat bed but would get much harder to use on a more complex geometry...

**Reply by wehrad** [reply]
Indeed when I look at nodesets only, the decoupling area still has the `bottom` nodeset:
![image](https://github.com/user-attachments/assets/4829dea2-e8c0-4c0e-8007-c4667739d467)
while having the `nodelist_6` too: 
![image](https://github.com/user-attachments/assets/548ca535-8ba6-4988-9ddb-a7ecaf86a27a)

So I guess there's some nodeset priority happening in the background, where the bottom nodeset takes over when the decoupling nodeset is left free.

I'll try and find a way to get the replacement to really work, both on sidesets and nodesets.

**Comment by wehrad** [comment]
I face the same problem of non-free BC with `ParsedGenerateSideset`..

**Comment by wehrad** [comment]
After some more testing it really feels like `replace = true` of `SideSetsAroundSubdomainGenerator` doesn't actually replace the sideset (nor the nodeset)

EDIT: it feels like it was the mesh refinement that prevented the sideset to be removed. If I remove the mesh refinement then sideset get replaced, but still not nodesets...

I was looking for a replace in `NodeSetsFromSideSetsGenerator` but doesn't seem like the way to go.

**Reply by GiudGiud** [reply]
even if it does not work, I think there is a way forward:

- create the limited sideset decoupling_bottom
- delete bottom using a BoundaryDeletionGenerator
- re-add bottom using a ParsedGenerateSideset, using the ParseExpression to exclude decoupling_bottom

I think we need in ParsedGenerateSideset an "exclude_boundaries" parameter. I can work on this soon. This will make creating bottom while excluding decoupling_bottom easier

**Reply by GiudGiud** [reply]
you can try new option for excluding a boundary from consideration when making a new boundary here
https://github.com/idaholab/moose/pull/28550
